src/main/java/seedu/address/logic/commands/commandDelete/DeleteFinanceCommand.java

```diff
@@ -54,20 +54,41 @@ public Index getIndex(List<Finance> lastShownList) throws CommandException {
   protected void preprocessUndoableCommand(Model model) throws CommandException {
     List<Finance> lastShownList = model.getFilteredFinanceList();
     if (this.toDelete == null) {
-      if (targetIndex.getZeroBased() >= lastShownList.size()) {
-        throw new CommandException(Messages.MESSAGE_INVALID_FINANCE_DISPLAYED_INDEX);
+      if (!ID.isValidId(targetID.toString())) {
+        throw new CommandException(Messages.MESSAGE_INVALID_FINANCE_DISPLAYED_ID);
       }
-
-      this.toDelete = lastShownList.get(targetIndex.getZeroBased());
+      this.toDelete = getFinance(lastShownList);
+    }
+    if (this.targetID == null) {
+      this.targetID = getID(lastShownList);
     }
-
     if (this.targetIndex == null) {
       this.targetIndex = getIndex(lastShownList);
     }
   }
 
+  // Find way to abstract this
```

Comment:
Dat can you refer to my AssignStudentToCourseCommand, there i have some methods to check if a course contains a student etcetc, should be able to help you abstract here